pkg/kube/krt/collection.go

```diff
@@ -0,0 +1,383 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package krt
+
+import (
+	"fmt"
+	"sync"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"istio.io/istio/pkg/kube/controllers"
+	istiolog "istio.io/istio/pkg/log"
+	"istio.io/istio/pkg/maps"
+	"istio.io/istio/pkg/ptr"
+	"istio.io/istio/pkg/slices"
+	"istio.io/istio/pkg/util/sets"
+)
+
+type manyCollection[I, O any] struct {
+	parent Collection[I]
+	log    *istiolog.Scope
+
+	// mu protects all items grouped below
+	mu              sync.Mutex
+	collectionState multiIndex[I, O]
+	dependencies    sets.String
+	// Stores a map of I -> Things depending on it
+	objectRelations map[Key[I]]dependencies
+
+	handlersMu sync.RWMutex
+	handlers   []func(o []Event[O])
+
+	handle HandleMulti[I, O]
+}
+
+type multiIndex[I, O any] struct {
+	objects  map[Key[O]]O
+	inputs   map[Key[I]]sets.Set[Key[O]]
+	inputObj map[Key[I]]I
+}
+
+type dump interface {
+	Dump()
+}
+
+func Dump[O any](c Collection[O]) {
+	if d, ok := c.(dump); ok {
+		d.Dump()
+	} else {
+		log.Warnf("cannot dump collection %T", c)
+	}
+}
+
+func (h *manyCollection[I, O]) Dump() {
+	h.mu.Lock()
+	defer h.mu.Unlock()
+	h.log.Errorf(">>> BEGIN DUMP")
+	for k, v := range h.objectRelations {
+		for kk, vv := range v.dependencies {
+			h.log.Errorf("Dependencies for: %v: %v -> %v (%v)", k, kk, vv.key, vv.filter)
+		}
+	}
+	for i, os := range h.collectionState.inputs {
+		h.log.Errorf("Input %v -> %v", i, os.UnsortedList())
+	}
+	for os, o := range h.collectionState.objects {
+		h.log.Errorf("Output %v -> %v", os, o)
+	}
+	h.log.Errorf("<<< END DUMP")
+}
+
+// onUpdate takes a list of I's that changed and reruns the handler over them.
+func (h *manyCollection[I, O]) onUpdate(items []Event[any]) {
+	var events []Event[O]
+	for _, a := range items {
+		i := a.Latest().(I)
+
+		iKey := GetKey(i)
+		log := h.log.WithLabels("key", iKey)
+
+		h.mu.Lock()
+		// Find dependencies for this input; insert empty set if we don't have any
+		d, f := h.objectRelations[iKey]
+		if !f {
+			d = dependencies{
+				dependencies: map[depKey]dependency{},
+				finalized:    true, // TODO: set this to true at some point
```

Comment:
The comment implies that finalized is meant to be initialized as false. I think the comment can just be removed though because this should be handling the case where there wasn't a know relation to I